src/compute-client/src/controller/instance.rs

```diff
@@ -629,31 +598,94 @@ impl<T: ComputeControllerTimestamp> Instance<T> {
 
                 let key = (histogram_period, bucket, histogram_labels.clone());
                 *stash.entry(key).or_default() += Diff::ONE;
+            }
+        }
+    }
 
-                if refresh_histogram {
-                    for ((period, lag, labels), count) in std::mem::take(stash) {
-                        let mut packer = row_buf.packer();
-                        packer.extend([
-                            Datum::TimestampTz(period.start),
-                            Datum::TimestampTz(period.end),
-                            Datum::String(&collection_id.to_string()),
-                            Datum::UInt64(lag),
-                        ]);
-                        let labels = labels.iter().map(|(k, v)| (*k, Datum::String(v)));
-                        packer.push_dict(labels);
-
-                        histogram_updates.push((row_buf.clone(), count));
-                    }
-                }
+    /// Produce new wallclock lag introspection updates, provided enough time has passed since the
+    /// last recording.
+    //
+    /// We emit new introspection updates if the system time has passed into a new multiple of the
+    /// recording interval (typically 1 minute) since the last refresh. The storage controller uses
+    /// the same approach, ensuring that both controllers commit their lags at roughly the same
+    /// time, avoiding confusion caused by inconsistencies.
+    fn maybe_record_wallclock_lag(&mut self) {
+        if self.read_only {
+            return;
+        }
+
+        let duration_trunc = |datetime: DateTime<_>, interval| {
+            let td = TimeDelta::from_std(interval).ok()?;
+            datetime.duration_trunc(td).ok()
+        };
+
+        let interval = WALLCLOCK_LAG_RECORDING_INTERVAL.get(&self.dyncfg);
+        let now_dt = mz_ore::now::to_datetime((self.now)());
+        let now_trunc = duration_trunc(now_dt, interval).unwrap_or_else(|| {
+            soft_panic_or_log!("excessive wallclock lag recording interval: {interval:?}");
+            let default = WALLCLOCK_LAG_RECORDING_INTERVAL.default();
+            duration_trunc(now_dt, *default).unwrap()
+        });
+        if now_trunc <= self.wallclock_lag_last_recorded {
+            return;
+        }
```

Comment:
Does this risk causing workload spikes at the minute interval? It seems the previous approach smeared the work across the interval, while the new approach piles all updates on the same moment in time. This works as long as the amount of updates doesn't grow too much, but can cause problems once we track the lag of a significant number of collections.